### PR TITLE
[feat]tools-v2: add check copyset

### PR DESCRIPTION
**Pull request by baytan0720**
### What problem does this PR solve?

Issue Number: #2354 

Problem Summary: support check copyset command in [curve tool](https://github.com/opencurve/curve/tree/master/tools-v2)

### What is changed and how it works?

What's Changed:
add tools-v2/pkg/cli/command/curvebs/check/check.go
add tools-v2/pkg/cli/command/curvebs/check/copyset/copyset.go
add tools-v2/pkg/cli/command/curvebs/status/copyset/get_copyset_status.go
modify tools-v2/pkg/cli/command/curvebs
modify tools-v2/pkg/config/bs.go
modify tools-v2/internal/utils/copyset.go
modify tools-v2/internal/error/error.go
modify tools-v2/README.md

How it Works:

```shell
root@09e6445fe6e0:/# curve bs check copyset -h                                
Usage:  curve bs check copyset [flags]

check copysets health in curvebs

Flags:
  -c, --conf string             config file (default is
                                $HOME/.curve/curve.yaml or
                                /etc/curve/curve.yaml)
      --copysetid strings       copyset ids[required]
  -f, --format string           output format (json|plain) (default "plain")
  -h, --help                    print help
      --logicalpoolid strings   copyset ids[required]
      --margin uint             the maximum gap between peers (default 1000)
      --mdsaddr string          mds address, should be like
                                127.0.0.1:6700,127.0.0.1:6701,127.0.0.1:6702
      --rpcretrytimes int32     rpc retry times (default 1)
      --rpctimeout duration     rpc timeout (default 10s)
      --showerror               display all errors in command
      --verbose                 show some log

Examples:
$ curve bs check copyset --copysetid 1 --logicalpoolid 1
```

command usage

```shell
root@09e6445fe6e0:/# curve bs check copyset --copysetid 1 --logicalpoolid 1
+------------+-----------+--------+--------+--------+---------+
| COPYSETKEY | COPYSETID | POOLID | STATUS | LOGGAP | EXPLAIN |
+------------+-----------+--------+--------+--------+---------+
| 4294967297 | 1         | 1      | ok     | 0      |         |
+------------+-----------+--------+--------+--------+---------+
root@09e6445fe6e0:/# curve bs check copyset --copysetid 1,2 --logicalpoolid 1,1
+------------+-----------+--------+--------+--------+---------+
| COPYSETKEY | COPYSETID | POOLID | STATUS | LOGGAP | EXPLAIN |
+------------+-----------+--------+--------+--------+---------+
| 4294967297 | 1         | 1      | ok     | 0      |         |
+------------+-----------+        +        +--------+---------+
| 4294967298 | 2         |        |        | 0      |         |
+------------+-----------+--------+--------+--------+---------+
```

## 

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [x] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by baytan0720** [comment]
cicheck

File: tools-v2/pkg/cli/command/curvebs/check/copyset/gap.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+// known_applied_index: 3
+// last_log_id: (index=3,term=3)
+// state_machine: Idle
+// last_committed_index: 3
+// last_snapshot_index: 3
+// last_snapshot_term: 3
+// snapshot_status: IDLE
+// replicator_3298534883333@10.219.192.50:6800:0: next_index=4  flying_append_entries_size=0 idle hc=150785 ac=2 ic=0
+// replicator_2203318222854@10.219.192.50:6802:0: next_index=4  flying_append_entries_size=0 idle hc=150785 ac=2 ic=0
+func GetLeaderCopysetGap(addr string, key2LeaderInfo *sync.Map, timeout time.Duration) *cmderror.CmdError {
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/check/copyset/gap.go:113]: just use fs same func is enough?

File: tools-v2/pkg/cli/command/curvebs/check/copyset/status.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
+/*
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/check/copyset/status.go:1]: Can this be packaged into a command like [chunkserver](https://github.com/opencurve/curve/blob/44b695e7ed9f83144f19712e804dc56899ce2e5d/tools-v2/pkg/cli/command/curvebs/query/chunk/chunkserver.go) ?Maybe this command is only called by other commands and does not need to be added. Once it is added, it is convenient to modify.

File: tools-v2/pkg/cli/command/curvebs/check/copyset/copyset.go [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+func GetCopysetsStatus(caller *cobra.Command, copysetIds string, poolIds string) (interface{}, *tablewriter.Table, *cmderror.CmdError, cobrautil.ClUSTER_HEALTH_STATUS) {
+	checkCopyset := NewCheckCopysetCommand()
+	checkCopyset.Cmd.SetArgs([]string{
+		fmt.Sprintf("--%s", config.CURVEFS_COPYSETID), copysetIds,
+		fmt.Sprintf("--%s", config.CURVEFS_POOLID), poolIds,
+		fmt.Sprintf("--%s", config.FORMAT), config.FORMAT_NOOUT,
+	})
+	config.AlignFlagsValue(caller, checkCopyset.Cmd, []string{
+		config.RPCRETRYTIMES, config.RPCTIMEOUT, config.CURVEFS_MDSADDR,
+		config.CURVEFS_MARGIN,
```
[the remark this answers is not in the document]
Comment (baytan0720) [reply tools-v2/pkg/cli/command/curvebs/check/copyset/copyset.go]: > should be `config.CURVEBS_MDSAADR` and `config.CURVEBS_MARGIN`? Though the underlying strings are same.

fix

**Comment by baytan0720** [comment]
cicheck

**Comment by baytan0720** [comment]
cicheck

File: tools-v2/pkg/cli/command/curvebs/check/copyset/copyset.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+	config.AddBSCopysetIdSliceRequiredFlag(cCmd.Cmd)
+	config.AddBSLogicalPoolIdSliceRequiredFlag(cCmd.Cmd)
+	config.AddBsMarginOptionFlag(cCmd.Cmd)
+}
+
+func (cCmd *CopysetCommand) Init(cmd *cobra.Command, args []string) error {
+	mdsAddrs := config.GetBsFlagString(cmd, config.CURVEBS_MDSADDR)
+	retrytime := config.GetBsFlagInt32(cmd, config.RPCRETRYTIMES)
+	timeout := config.GetFlagDuration(cCmd.Cmd, config.RPCTIMEOUT)
+	margin := config.GetMarginOptionFlag(cCmd.Cmd)
```
Comment (Cyber-SiKu) [inline tools-v2/pkg/cli/command/curvebs/check/copyset/copyset.go:92]: GetBsMargin?

**Comment by baytan0720** [comment]
cicheck

File: tools-v2/pkg/cli/command/curvebs/check/copyset/copyset.go (hunk 4 above)
Comment (baytan0720) [reply tools-v2/pkg/cli/command/curvebs/check/copyset/copyset.go]: fix all

**Comment by baytan0720** [comment]
cicheck

**Comment by baytan0720** [comment]
cicheck

**Review by Cyber-SiKu — APPROVED** [review]

**Comment by Cyber-SiKu** [comment]
cicheck

**Review by wu-hanqing — APPROVED** [review]

**Pull request merged**